Framework integration

jbolns opened the pull request:
Not ready to merge. Simply draft PR-ing for discussion.

purva-thakre commented on docs/gallery/using_nisq_circuits.ipynb:
```diff
@@ -0,0 +1,530 @@
+    "import random\n",
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "%matplotlib widget\n",
```
right now the docs build is failing because it does not like this line.

jbolns replied:
Commented it out. If running locally, this is needed for interactive 3D visualisations, so don't want to completely delete. But it makes sense that it won't run in the docs, as, even locally, it requires authorising the widget download.

purva-thakre commented on docs/gallery/using_nisq_circuits.ipynb:
```diff
@@ -0,0 +1,530 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from topologiq.scripts.runner import runner\n",
+    "from topologiq.utils.interop_pyzx import pyzx_g_to_simple_g\n",
```
`build-and-test-python` workflows are failing because `uv` cannot find the `topologiq` installation. Looks like topologiq should be a tests dependency as well as docs dependency in the toml file.

purva-thakre commented on docs/gallery/using_nisq_circuits.ipynb:
```diff
@@ -0,0 +1,530 @@
+   "id": "de2c9b1d",
+   "metadata": {},
+   "source": [
+    "The incoming circuit is not optimised. Topologiq can handle an unoptimised circuit, but the space-time volume of the diagram produced by topologiq is increasingly proportional to the size of the input graph. It is therefore a good idea to optimise the incorming circuit. \n",
```
```suggestion
    "The incoming circuit is not optimised. Topologiq can handle an unoptimised circuit, but the space-time volume of the diagram produced by topologiq is increasingly proportional to the size of the input graph. It is therefore a good idea to optimise the incoming circuit. \n",
```

purva-thakre commented on docs/gallery/using_nisq_circuits.ipynb:
```diff
@@ -0,0 +1,530 @@
+    "zx_graph.apply_state('0' * num_apply_state)\n",
+    "\n",
+    "# Apply post-select only to the outputs of the ancilla qubits\n",
+    "# Credits. C, Purva Thakre\n",
```
```suggestion

```
no need for this. Thanks for being thoughtful!

purva-thakre commented on docs/gallery/using_nisq_circuits.ipynb:
```diff
@@ -0,0 +1,530 @@
+    "            decoders=[\"pymatching\"],\n",
+    "            print_progress=False,\n",
+    "            save_resume_filepath=Path(\n",
+    "                f\"../_examples_database/steane_stats_{support_observable_basis.value}.csv\"\n",
```
what are your thoughts on renaming this? There already exists a database with this name due to `docs/gallery/steane_encoding.ipynb`

purva-thakre commented on docs/gallery/using_nisq_circuits.ipynb:
```diff
@@ -0,0 +1,530 @@
+    "                / f\"logical_cnot_result_{support_observable_basis}_observable_{i}.png\"\n",
+    "            )\n",
+    "\n",
+    "# Function to cal simulation cycle for all relevant bases\n",
```
```suggestion
    "# Function to call the simulation cycle for all relevant bases\n",
```

afowler commented:
My surprise was actually about cirq, not qiskit, since I thought cirq was
required for stim. It's possible this tool simply gets installed when you
installed stim.

On Tue, Oct 7, 2025, 6:10 AM J ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/tqec/interop/nisq/load.py
> <https://github.com/tqec/tqec/pull/720#discussion_r2410583448>:
>
> > +    qc.h(0)
> +    qc.cx(0, 3)
> +    qc.cx(0, 4)
> +    qc.cx(0, 5)
> +    qc.cx(0, 6)
> +    qc.h(0)
> +
> +    qc.h(1)
> +    qc.cx(1, 3)
> +    qc.cx(1, 4)
> +    qc.cx(1, 7)
> +    qc.cx(1, 8)
> +    qc.h(1)
> +
> +    qc.h(2)
> +    qc.cx(2, 3)
> +    qc.cx(2, 5)
> +    qc.cx(2, 7)
> +    qc.cx(2, 9)
> +    qc.h(2)
>
> Yes, we should add it as .qpy, steane.qpy should do it. Do you want to do
> it or should I do it? I'm fine either way.
>
> In a conversation with @afowler <https://github.com/afowler> we were
> surprised that qiskit is not a default dependency already. It used to be,
> no? Since it's the major such framework, it makes sense to ship it default.
> At the very least, it keeps our own versioning in check.
>
> As the notebook runs qiskit by default, this suffices to avoid runtime
> errors.
>
> Other frameworks... Let's pin that until after the next push in this draft
> PR. I have a suggestion, but best not to mix topics.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tqec/tqec/pull/720#discussion_r2410583448>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAXTF5EAU2X5G5Y2HJHK33WO3TTAVCNFSM6AAAAACIISKLYCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTGMJQGA2DQMJUHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jbolns commented:
> cirq, not qiskit
My bad, I mixed them up. 

Cirq is not installing at all, even after my last push containing additional dependencies. 

I was going to add it explicitly but:
- we don't yet have a cirq steane (in that conversation I mixed up acronyms, we have qiskit and qrisp)
- there's a versioning problem with cirq that would require loosening TQEC's main Python requirements (to include 3.10) and I can't immediately tell if this could have collateral effects.

jbolns commented:
So, the latest push has a few things worth noting explicitly. 

**Dependencies**
I added a new dependency group to `pyproject.toml`. 
- A foundational install wouldn't install topologiq, nor qiskit, nor qrisp, not any other framework we might add in the future. 
- If you add the `docs` dependencies, it would install all pre-existing doc dependecies plus topologiq and qiskit because that's all the notebook needs to run in its default configuration
- There is an additional `integration` group with topologiq and all supported frameworks, which is what you would install if running comprehensive integration tests locally (or, more practically, you would `uv add <applicable_framework>` only the framework you're interested in). 

The notebook contains instructions to clarify all this. It's easier to explain for an user than to give the full summary above, as they just need to know they need to run the group or add their chosen framework library.  

**Notebook & circuits**
Additionally:
- Uploaded the .qpy Yilun linked above to `./assets`
- Updated the nisq importer and notebook to reflect this
- Updated notebook to consider all feedback in this thread
- Updated notebook to consider above-noted changes to dependencies.

purva-thakre commented:
> My surprise was actually about cirq, not qiskit, since I thought cirq was
required for stim.

@afowler There is now a separate package called [stimcirq](https://pypi.org/project/stimcirq/). We only install Stim without the additional package.

purva-thakre commented:
@jbolns Now, the docs build is failing because your notebook needs to be added to `docs/gallery/index.rst`.

purva-thakre commented on pyproject.toml:
```diff
@@ -81,6 +81,8 @@ docs = [
   "sphinx-copybutton<=0.5.2",
   "sphinxcontrib-bibtex<=2.64",
   "jupyter-sphinx<=0.5.3",
+  "qiskit>=2.1.1",
+  "topologiq@git+https://github.com/tqec/topologiq.git"
```
since you added the integration group, you can directly use it here https://packaging.python.org/en/latest/specifications/dependency-groups/

purva-thakre replied:
See the example at the provided link:

<img width="833" height="414" alt="Image" src="https://github.com/user-attachments/assets/28fd710f-7b09-4532-94c1-a46398573dd3" />

jbolns commented on src/tqec/interop/nisq/load.py:
```diff
@@ -0,0 +1,134 @@
+    qc.cx(2, 5)
+    qc.cx(2, 7)
+    qc.cx(2, 9)
+    qc.h(2)
```
Latest push solves this, I think. It's now defined in Python, as .py, and simply imported into the notebook. See comment below about potentially needing separate doc files if we add more examples (in which case, perhaps qrisp would be a good candidate for a notebook of its own).

jbolns replied:
I think this one is resolved.

afowler commented:
I think it is okay for a user to focus on a single framework and do a
single set of sims per framework. It is not obvious that two different
frameworks even with in principle the same quantum circuit will lead to
exactly the same lattice surgery and therefore exactly the same simulations.

On Wed, Oct 8, 2025, 3:43 AM J ***@***.***> wrote:

> *jbolns* left a comment (tqec/tqec#720)
> <https://github.com/tqec/tqec/pull/720#issuecomment-3380918000>
>
> With the exception of the comment I noted explicitly not understanding,
> the latest push should address all comments thus far.
>
> *Changelog summary*
> The most significant change is that, having externalised the circuits to a
> corresponding file in ./assets/, it is now possible to import them
> directly into the notebook.
>
> As a consequence, I went ahead and completely eliminated the
> ./src/tqec/interops/nisq I had introduced earlier, as it was no longer
> being used by the notebook.
>
> *A few additional notes*
> While I was very sold on the idea of a single notebook for many
> frameworks, I tried adding a cirq version using the circuit Angela uploaded
> to the algorithm's folder. At that point, the notebook did become a bit too
> long. I think two frameworks fit perfectly into a single notebook, but if
> we are going to add examples of more frameworks, we might want to consider
> several framework-specific notebooks. Problem is, this would significantly
> add simulations to the deployment workflow, one full set of sims per
> notebook.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tqec/tqec/pull/720#issuecomment-3380918000>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAXTBJUH267A4KNNF3WHT3WTTGDAVCNFSM6AAAAACIISKLYCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGOBQHEYTQMBQGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

purva-thakre commented on assets/steane_qrisp.py:
```diff
@@ -0,0 +1,48 @@
+# The following is how you would generate a Steane code using Qrisp.
```
Please add some comments along the lines of this file is used by so and so notebook in the documentation.

purva-thakre commented on docs/integration/index.rst:
```diff
@@ -0,0 +1,13 @@
+
+This section contains examples of how to use (NISQ) circuits designed with external frameworks including, but not limited to, Qiskit and Cirq.
+
+.. nbintegration::
```
You also need to add the new `index.rst` file to `docs/gallery/index.rst` if you want to list the integration examples on the gallery page.

jbolns replied:
Oh, so that's a pre-established tag. I thought it was the section tile. I really need to learn some Sphinx syntax. 🤣
- Changed line. 
- Added integration gallery to docs index, to see if it displays as full section. I think that's what @inmzhang meant by separate gallery (which makes sense if we only use a single notebook, else we put back to main gallery if there's only one integration notebook).

purva-thakre replied:
> I think that's what @inmzhang meant by separate gallery (which makes sense if we only use a single notebook

I am confused. I though we were adding a new section to the examples gallery. See below:

<img width="571" height="603" alt="image" src="https://github.com/user-attachments/assets/ab94267c-155b-47a0-b7b2-ecf52da07169" />

jbolns replied:
That's exactly what I tried to achieve. 🤣
Not sure if it's been evident, but I'm not particularly getting the whole Sphinx docs rationale. 

Will push commits with all the other fixes but, to avoid trying recipes blind, could you clarify the structure for me. 

Considering the two structural rationales I've pushed are not the intended outcome, the next approach I would try would be: 
- move `qiskit.ipynb` and `qrisp.ipynb` notebooks inside `./docs/gallery`
- edit `docs/gallery/index.rst` as follows

```
This section contains a collection of constructed logical computations and their simulation results.

.. nbgallery::

   memory.ipynb
   cnot.ipynb
   move_rotation.ipynb
   three_cnots.ipynb
   steane_encoding.ipynb

.. nbgallery::

   qiskit.ipynb
   qrisp.ipynb

```

Would this be correct, i.e., lead to what this picture just above shows?

jbolns commented:
**Summary of last push.**
- Improves dependency management
- Improves readability of notebooks
- Eliminates ty failures
- Realises the multi-gallery approach: standard gallery + integration gallery. 

**Pending**
Per my count, the pending issues to discuss are:
- Optimising tests introduced in #726 by using parametrize to condense multiple unit tests like this into a single unit test
- Reducing docstrings of `.../interops/pyzx/topologiq.py` bug-fix introduced in #726 to the standard TQEC length and style
- Deciding between ONE or MANY notebooks approach, alongside their placement. 

**One or many?**

> I think it is okay for a user to focus on a single framework and do a single set of sims per framework. It is not obvious that two different frameworks even with in principle the same quantum circuit will lead to exactly the same lattice surgery and therefore exactly the same simulations.

I think @afowler meant it is preferable to use several notebooks focusing each on a single framework. This avoids a Qiskit user having to worry about Qrisp/pytket/Cirq code and vice-versa. 

> One notebook only for framework integration, where the code randomly selects which framework's circuit to use

This is similar to my initial idea of letting users select a framework, with the improvement that this would automatically run different frameworks every time there's a merge. So, I can't say I fundamentally disagree. It would be fine. 

I'm siding with different notebook for each framework tested successfully because they're easier to read by users. 

Having said that, I would prefer if the designers of the circuits for this or that framework prepare and push their respective notebooks. Not too worried about the Qiskit and Qrisp notebooks since both @purva-thakre and @Zhaoyilunnn are established contributors. But newer contributors who go over the pain of designing a circuit and testing it should get the lines/credit.

purva-thakre commented on docs/integration/qiskit.ipynb:
```diff
@@ -0,0 +1,574 @@
+    "\n",
+    "The notebook assumes you have already designed a Qiskit circuit and saved it using Qiskit's `.qpy` format. \n",
+    "\n",
+    "> **NB!** Run `uv add --only-group integration` to install the libraries used in this notebook (ignore if you installed all dependency groups when setting up TQEC)."
```
You cannot use this in sphinx. 

```suggestion
    .. note::
        Run `uv add --only-group integration` to install the libraries used in this notebook (ignore if you installed all dependency groups when setting up TQEC)."
```

purva-thakre commented:
> Having said that, I would prefer if the designers of the circuits for this or that framework prepare and push their respective notebooks. 

If you have access to their notebooks, you can add them as co-authors in github. https://gist.github.com/lisawolderiksen/f9747a3ae1e58e9daa7d176ab98f1bad

purva-thakre commented on docs/integration/qrisp.ipynb:
```diff
@@ -0,0 +1,581 @@
+    "\n",
+    "The notebook assumes you have already designed a Qrisp circuit. \n",
+    "\n",
+    "> **NB!** Run `uv add --only-group integration` to install the libraries used in this notebook (ignore if you installed all dependency groups when setting up TQEC)."
```
Apply https://github.com/tqec/tqec/pull/720#discussion_r2434133702 here.

purva-thakre commented on docs/integration/qiskit.ipynb:
```diff
@@ -0,0 +1,574 @@
+   "outputs": [],
+   "source": [
+    "print(f\"\\n=> Converted Qiskit circuit of name {circuit_name.capitalize()} into the following QASM string:\\n\")\n",
+    "print(qasm_str) if qasm_str else print(\"ERROR. Failed to convert circuit to QASM\")"
```
```suggestion

```
Same question as https://github.com/tqec/tqec/pull/720/files#r2434193440

purva-thakre commented on docs/integration/qiskit.ipynb:
```diff
@@ -0,0 +1,574 @@
+    "\n",
+    "There is a need to perform some re-scaling and re-indexing of edges. In TQEC's `block_graph`, pipes occupy no space, which leads to adjacent cubes being closer to one another than in a space-time diagram where pipes do occupy space for the purposes of visualising things clearly.\n",
+    "\n",
+    "Fortunately, TQEC already has a way to import 3D objects like `lattice_nodes` and `lattice_edges`. TQEC can import 3D models encoded in COLLADA format (.dae), which use the same kind of coordinates and artifacts as `lattice_nodes` and `lattice_edges`. Accordingly, the following block of code relies on TQEC's COLLADA interoperability to realise similar transformations for `lattice_nodes` and  ...
```
```suggestion
    "Fortunately, TQEC already has a way to import 3D objects like `lattice_nodes` and `lattice_edges`. TQEC can import 3D models encoded in COLLADA format (`.dae`), which use the same kind of coordinates and artifacts as `lattice_nodes` and `lattice_edges`. Accordingly, the following block of code relies on TQEC's COLLADA interoperability to realise similar transformations for `lattice_nodes` and `lattice_edges`, leading into a `block_graph`."
```

purva-thakre commented:
Out of curiosity, what further work is needed to reduce the circuit volume in topologiq? I am simply comparing Austin's manually built structure in his slides with the [one from topologiq](https://docs.google.com/presentation/d/1RcMXB3kWtWWyM_LOQLkxo19Zaq4atLaV1fCsWqd78PQ/edit?slide=id.g39455a4c89c_0_581#slide=id.g39455a4c89c_0_581).

purva-thakre commented:
Something is not working as expected in the build. Will dig through the code later today. The cache hit is finding older caches. 

I tried triggering the build without the cache, and it took shorter than it should. My experience with the Steane simulation is that it should take the longest without the cache locally or in the CI. In this scenario, we have 3 notebooks for the Steane code that should have dominated the build time.

purva-thakre commented:
> The switch command would become larger and larger as we add frameworks, to the point the reader might get lost in the implementation of the switch rather than be able to focus in the specifics of the circuit.
> Alternative. How about we do different circuits? Steane for some notebooks, a CNOT for another, a few CNOTs for another.

@jbolns I will let you decide what you want to do. I do not think it will be that bad. Indeed, logical CNOT does not take that long, but it is not that interesting compared to the Steane code. 

Have some statements along the lines of: This notebook is compatible with multiple frameworks: Qiskit, Qrisp, etc. However, to reduce the simulation time, we demo the full pipeline in a randomly selected framework.

```py
import random

circuits = {'Qrisp': qrisp_circuit, 'Qiskit': qiskit_circuit}
selected_name = random.choice(list(circuits.keys()))
selected_circuit = circuits[selected_name]
print(f"{selected_name} circuit was selected")
```
Then, based on the value for `selected_name`, call the method that converts the circuit to QASM. Here, again, clarify why we are doing this. If this code block becomes too large, you can hide the code block in the jupyter notebook and allude to which method is used by which framework, etc. The rest of the pipeline is the same for all frameworks.

purva-thakre commented:
> The sky is the limit. If you want suggested ideas for improvement, I have at least 5. But I'm pretty sure there's many more than 5 ideas, so if you have ideas, don't be shy to open an issue.

thanks. I was simply curious as to why topologiq did not try to further reduce the circuit volume. I guess it isn't trying to create structures with the lowest possible circuit volume? If the tool does this, it will probably increase the run/execution time.

jbolns commented:
> Austin's slides for MQSF show topologiq constructed a 3D structure of volume 18, whereas the structure that Austin built by hand in SketchUp is of volume 12. 

Yes. I'm just interpreting the difference positively: Topologiq is doing such a good job reducing volumes that we're literally *only* 6 cubes from Austin's manual build.

Nonetheless, it remains true that it would be fantastic to get even lower volumes.

So, I'm glad you pressed on.

I just tried something and the **new best volume is now 14 cubes**.

Topologiq's 18-cube result corresponds to the graph in @afowler 's MQSF slide 16, i.e., exactly the graph PyZX outputs.

In slide 17, Austin adds a red spider to the ZX-graph before building manually.

If you feed the exact graph in slide 17 to Topologiq you get, amongst all variants, the following 14-cube output.

<img width="587" height="587" alt="steane_opt_output" src="https://github.com/user-attachments/assets/fa24ed16-c704-41a8-ad71-d05ba4513031" />

I suppose this means a worthy avenue for further research is to find a generalisable way to decide when and how to perform the kind of operation in slide 17. 

A good starting point for this would be to find the original algorithm proposal. It contained a similar operation, but we've lost the document and I can't recall the specifics.

Azuya334 commented:
@jbolns I made a integration notebook for qsim with the circuit defined in cirq. Should I make a branch and PR for it?

afowler commented:
I'll make revisiting this algorithm a priority when I get back. I can
re-create the lost document. Plus add a bunch of new ideas. Have added a
slide on the below to the presentation.

On Fri, Oct 17, 2025 at 1:00 AM J ***@***.***> wrote:

> *jbolns* left a comment (tqec/tqec#720)
> <https://github.com/tqec/tqec/pull/720#issuecomment-3414309117>
>
> Austin's slides for MQSF show topologiq constructed a 3D structure of
> volume 18, whereas the structure that Austin built by hand in SketchUp is
> of volume 12.
>
> Yes. I'm just interpreting the difference positively: Topologiq is doing
> such a good job reducing volumes that we're literally *only* 6 cubes from
> Austin's manual build.
>
> Nonetheless, it remains true that it would be fantastic to get even lower
> volumes.
>
> So, I'm glad you pressed on.
>
> I just tried something and the *new best volume is now 14 cubes*.
>
> Topologiq's 18-cube result corresponds to the graph in @afowler
> <https://github.com/afowler> 's MQSF slide 16, i.e., exactly the graph
> PyZX outputs.
>
> In slide 17, Austin adds a red spider to the ZX-graph before building
> manually.
>
> If you feed the exact graph in slide 17 to Topologiq you get, amongst all
> variants, the following 14-cube output.
> steane_opt_output.png (view on web)
> <https://github.com/user-attachments/assets/fa24ed16-c704-41a8-ad71-d05ba4513031>
>
> I suppose this means a worthy avenue for further research is to find a
> generalisable way to decide when and how to perform the kind of operation
> in slide 17.
>
> A good starting point for this would be to find the original algorithm
> proposal. It contained a similar operation, but we've lost the document and
> I can't recall the specifics.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tqec/tqec/pull/720#issuecomment-3414309117>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAXTDCQVWTDAQXCWSSY633YCOYFAVCNFSM6AAAAACIISKLYCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIMJUGMYDSMJRG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jbolns commented:
Super! 
Ps. Really like slides 27 & 28 (Y, S), too. #epic @inmzhang @KabirDubey !

jbolns commented:
> The only thing that is different across all frameworks is how the external frontend's circuit is lowered to QASM

After seeing examples of usage using other frameworks, having notebooks for each does seem helpful. Users generally prefer to see examples using their specific framework. Furthermore, not all frameworks are as similar to one another as Qiskit and Qrisp. In particular, some frameworks will altogether skip QASM.

> If we go down the path of one notebook per frontend, working on the docs locally will become quite the hassle.

For now, a compromise solution is to comment out the lines that trigger simulations in some of the notebooks. We can specify in the notebook that since the procedure to run a simulation is the same, to economise resources, we commented the call out. 

On the long-run, it would be ideal to get something like the examples_database work nicely for GH and local builds.

At the end of the day, our docs should scale alongside what we can do.

If we're having build-time issues with but-a-couple-notebooks, we need a different approach to docs.

purva-thakre commented:
> After seeing examples of usage using other frameworks, having notebooks for each does seem helpful.

Because their build times are not as bad as ours. 

> In particular, some frameworks will altogether skip QASM.

I am not sure what you mean by this. The current pipeline is literally external frontend -> qasm -> pyzx -> tqec. The only thing that is different across all is that the first step differs across all.  

> If we're having build-time issues with but-a-couple-notebooks, we need a different approach to docs.

It's not a couple notebooks. It is the entire gallery. As the examples become more complicated, simulation time takes over.

> On the long-run, it would be ideal to get something like the examples_database work nicely for GH and local builds.

The issue is also not just with this. There are things we can't control or further optimize, like Stim/sinter, which can take longer as the examples scale up. 

I think I am going to pull away from maintaining the docs after this. It is quite the headache making changes to the docs as the build time increases. I have tried to make my case, and the conversation is going nowhere.

smburdick commented:
> This PR is dated. Conversation on #862 is progressing under the radar. This PR can be closed Will give a couple days in case of last thoughts by anyone who has commented in it, then close.

@jbolns Are there any components you want to cherry pick from this, and submit as a PR? My suggestion would be to do the smaller parts individually, which makes them easier to review

jbolns commented:
@smburdick 
Hi man, thanks for asking. I'd forgotten to close issue.

Nothing here is both important and salvageable. Too old. Even the example notebooks would need to be rewritten. 

The replacement is very minimalist and we have reached an agreement on how to proceed with it, so will tag you there in case you want to be involved. Tomorrow, though. Want to do it from computer (now on mobile) to copy some code over.
